crates/sargon/src/profile/mfa/security_structures/automatic_shield_builder.rs

```diff
@@ -0,0 +1,528 @@
+        }
+    }
+
+    fn assign_factors_matching_selector(
+        &mut self,
+        to: RoleKind,
+        selector: FactorSelector,
+        quantity_to_add: Quantity,
+    ) {
+        let target_role = to;
+
+        let mut factors_to_add = self
+            .remaining_available_factors
+            .iter()
+            .filter(|&f| match selector {
+                FactorSelector::Category(category) => f.category() == category,
+                FactorSelector::Kind(kind) => f.factor_source_kind() == kind,
+            })
+            .map(|f| f.id())
+            .collect::<IndexSet<_>>();
+
+        if let Some(quantity) = quantity_to_add.as_fixed() {
+            factors_to_add = factors_to_add
+                .into_iter()
+                .take(*quantity)
+                .collect::<IndexSet<_>>();
+        }
+
+        self.remaining_available_factors
+            .retain(|f| !factors_to_add.contains(&f.id()));
+
+        self.proto_matrix
+            .add_factors_for_role(target_role, factors_to_add);
+    }
+
+    fn factors_for_role(&self, role: RoleKind) -> &IndexSet<FactorSourceID> {
+        self.proto_matrix.factors_for_role(role)
+    }
+
+    fn assign_factors_of_category(
+        &mut self,
+        to: RoleKind,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) {
+        self.assign_factors_matching_selector(
+            to,
+            FactorSelector::Category(category),
+            quantity_to_add,
+        )
+    }
+
+    fn assign_factors_of_kind(
+        &mut self,
+        to: RoleKind,
+        kind: FactorSourceKind,
+        quantity_to_add: Quantity,
+    ) {
+        self.assign_factors_matching_selector(
+            to,
+            FactorSelector::Kind(kind),
+            quantity_to_add,
+        )
+    }
+
+    fn assign_one_hardware_factor(&mut self, to: RoleKind) {
+        self.assign_factors_of_category(
+            to,
+            FactorSourceCategory::Hardware,
+            Quantity::one(),
+        )
+    }
+
+    fn count_factors_for_role(&self, role_kind: RoleKind) -> u8 {
+        self.factors_for_role(role_kind).len() as u8
+    }
+
+    fn assign_factors_of_category_to_recovery(
+        &mut self,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) {
+        self.assign_factors_of_category(Recovery, category, quantity_to_add)
+    }
+
+    fn assign_factors_of_category_to_confirmation(
+        &mut self,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) {
+        self.assign_factors_of_category(Confirmation, category, quantity_to_add)
+    }
+
+    /// Automatic assignment of factors to roles according to [this heuristics][doc].
+    ///
+    /// [doc]: https://radixdlt.atlassian.net/wiki/spaces/AT/pages/3758063620/MFA+Rules+for+Factors+and+Security+Shields#Automatic-Security-Shield-Construction
+    fn assign(&mut self) -> Result<ProtoMatrix> {
```

Comment:
@ymatto Would be great if you could take a look at my impl of the "Automatic Shield Construction" here.

We call `assign` on for a SecurityShieldBuilder which has already valid `primary_role threshold factors` set.